src/main/flight/alt_hold.c

```diff
@@ -0,0 +1,245 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Betaflight. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "platform.h"
+#include "alt_hold.h"
+
+#ifdef USE_ALTHOLD_MODE
+
+#include "drivers/time.h"
+#include "flight/failsafe.h"
+#include "flight/imu.h"
+#include "flight/position.h"
+#include "sensors/acceleration.h"
+#include "sensors/barometer.h"
+#include "config/config.h"
+#include "fc/runtime_config.h"
+#include "fc/rc.h"
+#include "osd/osd.h"
+#include "common/printf.h"
+#include "common/maths.h"
+#include "math.h"
+#include "build/debug.h"
+
+
+PG_REGISTER_WITH_RESET_TEMPLATE(altholdConfig_t, altholdConfig, PG_ALTHOLD_CONFIG, 3);
+
+PG_RESET_TEMPLATE(altholdConfig_t, altholdConfig,
+    .velPidP = 30,
+    .velPidD = 0,
+
+    .altPidP = 75,
+    .altPidI = 20,
+
+    .minThrottle = 6,
+    .maxThrottle = 65,
+);
+
+
+void simplePidInit(simplePid_s* simplePid, float min, float max, float kp, float kd, float ki)
+{
+    simplePid->max = max;
+    simplePid->min = min;
+    simplePid->kp = kp;
+    simplePid->kd = kd;
+    simplePid->ki = ki;
+    simplePid->lastErr = 0;
+    simplePid->integral = 0;
+}
+
+float simplePidCalculate(simplePid_s* simplePid, float dt, float targetValue, float measuredValue)
+{
+    float error = targetValue - measuredValue;
+
+    float pOut = simplePid->kp * error;
+
+    float iOut = simplePid->ki * simplePid->integral;
+
+    simplePid->integral += error * dt;
+
+    float derivative = (error - simplePid->lastErr) / dt;
+    float dOut = simplePid->kd * derivative;
+
+    float output = pOut + iOut + dOut;
+    output = constrainf(output, simplePid->min, simplePid->max);
+
+    simplePid->lastErr = error;
+    return output;
+}
+
+static float getCurrentAltitude(altHoldState_s* altHoldState)
+{
+#ifdef USE_BARO
+    if (sensors(SENSOR_BARO) && baroIsCalibrationComplete()) {
+        return 0.01f * baro.BaroAlt;
+    }
+#endif
+    float rawAltitude = 0.01f * getEstimatedAltitudeCm();
+    if (ABS(altHoldState->smoothedAltitude) < 0.01f) {
+        altHoldState->smoothedAltitude = rawAltitude;
+    }
+    float smoothFactor = 0.98f;
+    altHoldState->smoothedAltitude = (1.0f - smoothFactor) * rawAltitude + smoothFactor * altHoldState->smoothedAltitude;
+    return altHoldState->smoothedAltitude;
+}
+
+void altHoldReset(altHoldState_s* altHoldState)
+{
+    simplePidInit(&altHoldState->altPid, -50.0f, 50.0f,
+                  0.01f * altholdConfig()->altPidP,
+                  0.0f,
+                  0.01f * altholdConfig()->altPidI);
+
+    simplePidInit(&altHoldState->velPid, 0.0f, 1.0f,
+                  0.01f * altholdConfig()->velPidP,
+                  0.01f * altholdConfig()->velPidD,
+                  0.0f);
+
```

Comment:
Isn't velocity P the same as altitude D?
Maybe it could be easier to understand if we stick with the more common P, I and D labels for those three parameters?
Velocity D is the second derivative of altitude error, so it could be called Jerk or something like that and then we don't require two sets of PIDs, three for altitude, three for altitude velocity, where we don't use all three in either case?
Also I note that the default Altitude Velocity D (jerk) gain is zero, so perhaps we don't need it?

Reply:
Not sure here, but I assume that PID is taken from error variable, not controlled variable (compare with acro pids - input error is angle rate difference, output is motor thurst ~ angular moment)